[LISKDOCS-30] Add Velodrome (#253)
LiskHQ/lisk-docs: tutorials/cashback.md

This creates a simple Devnet, which is beneficial during development of the blockchain application.
The dummy delegates can be replaced by real delegates later on.

To start the network, execute the following command:

```bash
node index.js | npx bunyan -o short
```

`node index.js` will start the node, and `| npx bunyan -o short` will pretty-print the logs in the console.

Check the logs, to verify the network has started successfully.

If something went wrong, the process should stop and an error with debug information is displayed.

If everything is ok, you should be able to see similar logs like the ones displayed in [step 5 of the Hello World app tutorial](#5-start-the-network).

For further interaction with the network, you can run the process in the background by executing:

```bash
npx pm2 start --name cashback index.js # add the application to pm2 under the name 'cashback'
npx pm2 stop cashback # stop the cashback app
npx pm2 start cashback # start the cashback app
```

## 6. Interact with the network

Now that the network is started, let's try to send a `CashbackTransaction` to our node to see if it gets accepted.

As first step, create the transaction object.

> First, let's reuse the script [create_sendable_transaction.js](https://github.com/LiskHQ/lisk-sdk-examples/blob/development/scripts/create_sendable_transaction_base_trs.js) which we already described in [step 6 of Hello World app](#6-interact-with-the-network).

Reviewer: ```suggestion
First, let's reuse the script [create_sendable_transaction.js](https://github.com/LiskHQ/lisk-sdk-examples/blob/development/scripts/create_sendable_transaction_base_trs.js) which was created in step 6 of the other example application [Hello World](#6-interact-with-the-network).
```

A simple application which rewards its users for sending tokens. 

The Cashback App implementation goes as following:

- __Steps 1-5__ describe what needs to be implemented on the server-side of the blockchain application.
- __Step 6__ explains how to interact with the network from the client-side.

> Check out the __full code example__ for the [Cashback App on Github](https://github.com/LiskHQ/lisk-sdk-test-app/cashback).

## 1. Set up Lisk SDK

First, let's create the root folder for the Cashback App and initialize the project:

```bash
mkdir cashback # create the root folder for the blockchain application
cd cashback # navigate into the root folder
npm init # initialize the manifest file of the project
```

As next step, we want to install the `lisk-sdk` package and add it to our projects' dependencies.
Before installing it, make sure to follow the instructions in the [Lisk SDK - Pre-Install](../lisk-sdk/introduction.md#pre-installation) section.

```bash
npm install --save lisk-sdk@alpha # install lisk-sdk as dependency for the server side
npm install --save @liskhq/validator @liskhq/cryptography # install lisk-elements dependencies for the client side scripts
```

> Create 2 folders `client` and `server`, which will hold the corresponding scripts for the blockchain application.

Reviewer: ```suggestion
Create two folders `client` and `server`, which will hold the corresponding scripts for the blockchain application.
```

- the `server` folder holds all the code that is needed to start a node and connect it to a network.
- the `client` folder holds scritps that communicate with the network through the API of the node.

```bash

> mkdir client server # create the 2 folder inside the root directory of your blockchain application

Reviewer: ```suggestion
mkdir client server # creates the folder inside the root directory of your blockchain application
```

cd server # move inside the server folder
```

Inside the `server` folder, create the index file of your blockchain application:

```bash
touch index.js
```

## 2. Configure the application

> Next, let's configure the application, to provide basic information about the app we are going to build:

Reviewer: ```suggestion
Next, let's configure the application, to provide basic information about the application we are going to build:
```

```js
//server/index.js
const { Application, genesisBlockDevnet, configDevnet } = require('lisk-sdk'); // require application class, the default genesis block and the default config for the application

const app = new Application(genesisBlockDevnet, configDevnet); // create the application instance
```
> *See the complete file on Github: [cashback/server/index.js](https://github.com/LiskHQ/lisk-sdk-test-app/tree/development/cashback/server/index.js).*

> In the `line 2`, we require the needed dependencies from the `lisk-sdk` package.

Reviewer: ```suggestion
On `line 2`, we require the needed dependencies from the `lisk-sdk` package.
```

The most important one is the `Application` class, which is used in `line 4` to create the application instance.

> The application instance will start the whole application at the bottom of `index.js`.

Reviewer: ```suggestion
The application instance is used to start the blockchain application at the bottom of `index.js`.
```

In `line 4` , the application instance gets initialized.

> By passing the parameters for the [genesis block](../lisk-sdk/configuration.md) and the [configuration template](https://github.com/LiskHQ/lisk-sdk/blob/development/sdk/src/samples/config_devnet.json), the application is configured with most basic configurations to start the node.

Reviewer: ```suggestion
By passing the parameters for the [genesis block](../lisk-sdk/configuration.md) and the [configuration template](https://github.com/LiskHQ/lisk-sdk/blob/development/sdk/src/samples/config_devnet.json), the application is configured with the most basic configurations to start the node.
```

## 3. Create a new transaction type

Now, we want to create a new [custom transaction type](custom-transactions.md) `CashbackTransaction`: 
It extends the pre-existing transaction type `TransferTransaction`.

> In difference to the normal `TransferTransaction`, the Cashback transaction type pays out a 10% bonus reward to the sender of any Cashback transaction type.

Reviewer: ```suggestion
The difference between the regular `TransferTransaction` and the `CashbackTransaction`, is that Cashback transaction type also pays out a 10% bonus reward to its sender.
```

So e.g. if Alice sends 100 token to Bob as a Cashback transaction, Bob would receive the 100 token and Alice would receive additional 10 tokens as a cashback.

> > If you compare the methods below with the methods we implemented for the `HelloTransaction`, you will notice, that we implement less methods for the `CashbackTransaction`.

Reviewer: ```suggestion
> If you compare the methods below with the methods we implemented in the `HelloTransaction`, you will notice, that we implement fewer methods for the `CashbackTransaction`.
```

> This is because we extend the `CashbackTransaction` from an already existing transaction type `TransferTransaction`.

> > As a result, all required methods are implemented already inside the `Transfertransaction` class, and we only need to overwrite/extend explicitely the methods we want to customize.

Reviewer: ```suggestion
> As a result, all required methods are implemented already inside the `TransferTransaction` class, and we only need to overwrite/extend explicitely the methods we want to customize.
```

// the code block below starts the application and doesn't need to be changed
app
    .run()
    .then(() => app.logger.info('App started...'))
    .catch(error => {
        console.error('Faced error in application', error);
        process.exit(1);
    });
```

## 5. Start the network

Now, let's start our customized blockchain network for the first time.

The parameter `configDevnet`, which we pass to our `Application` instance in step 3, is preconfigured to start the node with a set of dummy delegates, that have enabled forging by default.
These dummy delegates stabilize the new network and make it possible to test out the basic functionality of the network with only one node immediately.

This creates a simple Devnet, which is beneficial during development of the blockchain application.
The dummy delegates can be replaced by real delegates later on.

To start the network, execute the following command:

```bash
node index.js | npx bunyan -o short
```

`node index.js` will start the node, and `| npx bunyan -o short` will pretty-print the logs in the console.

Check the logs, to verify the network has started successfully.

> If something went wrong, the process should stop and an error with debug information is displayed.

Reviewer: ```suggestion
If an error occurs, the process should stop and the error with debug information will be displayed.
```

> If everything is ok, you should be able to see the following console logs:

Reviewer: ```suggestion
If everything is ok, the following logs will be displayed:
```

14:01:39.800Z  INFO lisk-framework: App started...
14:01:39.818Z  INFO lisk-framework: Validating current block with height 1886
14:01:39.819Z  INFO lisk-framework: Loader->validateBlock Validating block 10258884836986606075 at height 1886
14:01:40.594Z  INFO lisk-framework: Lisk started: 0.0.0.0:4000
14:01:40.600Z  INFO lisk-framework: Verify->verifyBlock succeeded for block 10258884836986606075 at height 1886.
14:01:40.600Z  INFO lisk-framework: Loader->validateBlock Validating block succeed for 10258884836986606075 at height 1886.
14:01:40.600Z  INFO lisk-framework: Finished validating the chain. You are at height 1886.
14:01:40.601Z  INFO lisk-framework: Blockchain ready
14:01:40.602Z  INFO lisk-framework: Loading 101 delegates using encrypted passphrases from config
14:01:40.618Z  INFO lisk-framework: Forging enabled on account: 8273455169423958419L
14:01:40.621Z  INFO lisk-framework: Forging enabled on account: 12254605294831056546L
14:01:40.624Z  INFO lisk-framework: Forging enabled on account: 14018336151296112016L
14:01:40.627Z  INFO lisk-framework: Forging enabled on account: 2003981962043442425L
[...]
```

## 6. Interact with the network

> Now that the network is started, let's try to send a `CashbackTransaction` to our node to see if it gets accepted.

Reviewer: ```suggestion
Now that your network is running, let's try to send a `CashbackTransaction` to our node to see if it gets accepted.
```

As first step, create the transaction object.

> First, let's reuse the script [create_sendable_transaction_base_trs.js](https://github.com/LiskHQ/lisk-sdk-examples/blob/development/scripts/create_sendable_transaction_base_trs.js) which we already described in [step 6 of Hello World app](#6-interact-with-the-network).

Reviewer: ```suggestion
First, let's reuse the script [create_sendable_transaction_base_trs.js](https://github.com/LiskHQ/lisk-sdk-examples/blob/development/scripts/create_sendable_transaction_base_trs.js) which we already described in [step 6 of Hello World app](/hello-world.md#6-interact-with-the-network).
```
👆 Not sure if that link is correct, but I think it needs changing to point to the Hello World tutorial.

Author: @RachBLondon: This part of docs might be outdated soon, when this branch is merged: https://github.com/LiskHQ/lisk-sdk/compare/release/2.1.0...3739-create_sendable

I hope it gets merged today :)

> We can use this function in our script for printing a sendable `CashbackTransaction` object:

Reviewer: ```suggestion
We can call the `createSendableTransaction` function to print a sendable `CashbackTransaction` object:
```

 	passphrase: 'wagon stock borrow episode laundry kitten salute link globe zero feed marble', // the senders passphrase, needed to sign the transaction
 	secondPassphrase: null,
 	timestamp: 2,
});

console.log(c); // the transaction is displayed as JSON object in the console
```
> *See the complete file on Github: [hello_world/client/print_sendable_cashback.js](https://github.com/LiskHQ/lisk-sdk-examples/blob/development/hello_world/client/print_sendable_hello-world.js).*

The generated transaction object should look like this:
```json
{  
   "id":"5372254888441494149",
   "amount":"100000000",
   "type":11,
   "timestamp":3,
   "senderPublicKey":"c094ebee7ec0c50ebee32918655e089f6e1a604b83bcaa760293c61e0f18ab6f",
   "senderId":"16313739661670634666L",
   "recipientId":"10881167371402274308L",
   "fee":"10000000",
   "signature":"0a3f41cc529f9de523cadc7db64e9436014d1b10ca2158bbce0469e8e76dfd021358496440da43acaf64d0223d3514609fc1aa41646be56353207d88a03b1305",
   "signatures":[],
   "asset":{}
}
```

Now that we have a sendable transaction object, let's send it to our node and see how it gets processed by analyzing the logs.

> For this, we utilize the API of the node and post the created transaction object to the transaction endpoint of the API.

Reviewer: ```suggestion
For this, we utilize the http API of the node and post the created transaction object to the transaction endpoint of the API.
```
